hermes/src/store/proof/wormhole_merkle.rs

```diff
@@ -0,0 +1,148 @@
+use {
+    crate::store::{
+        types::MessageState,
+        AccumulatorState,
+        Store,
+    },
+    anyhow::{
+        anyhow,
+        Result,
+    },
+    byteorder::{
+        BigEndian,
+        WriteBytesExt,
+    },
+    pythnet_sdk::{
+        accumulators::{
+            merkle::{
+                MerkleAccumulator,
+                MerklePath,
+            },
+            Accumulator,
+        },
+        hashers::keccak256_160::Keccak160,
+    },
+    std::io::{
+        Cursor,
+        Write,
+    },
+};
+
+type Hash = [u8; 20];
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct WormholeMerkleProof {
+    pub vaa:         Vec<u8>,
+    pub state_index: u32,
+    pub root:        Hash,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct WormholeMerkleMessageProof {
+    pub vaa:   Vec<u8>,
+    pub proof: MerklePath<Keccak160>,
+}
+
+pub async fn store_wormhole_merkle_verified_message(
+    store: &Store,
+    proof: WormholeMerkleProof,
+) -> Result<()> {
+    let pending_acc = store
+        .pending_accumulations
+        .entry(proof.state_index)
+        .or_default()
+        .await
+        .into_value();
+    store
+        .pending_accumulations
+        .insert(proof.state_index, pending_acc.wormhole_merkle_proof(proof))
+        .await;
+    Ok(())
+}
+
+pub fn construct_message_states_proofs(
+    state: AccumulatorState,
+) -> Result<Vec<WormholeMerkleMessageProof>> {
+    // Check whether the state is valid
+    let merkle_acc = match MerkleAccumulator::<Keccak160>::from_set(
+        state
+            .accumulator_messages
+            .messages
+            .iter()
+            .map(|m| m.as_ref()),
+    ) {
+        Some(merkle_acc) => merkle_acc,
+        None => return Ok(vec![]), // It only happens when the message set is empty
+    };
+
+    let proof = &state.wormhole_merkle_proof;
+
+    log::info!(
+        "Merkle root: {:?}, Verified root: {:?}",
+        merkle_acc.root,
+        proof.root
+    );
+    log::info!("Valid: {}", merkle_acc.root == proof.root);
+
+    state
+        .accumulator_messages
+        .messages
+        .iter()
+        .map(|m| {
+            Ok(WormholeMerkleMessageProof {
+                vaa:   state.wormhole_merkle_proof.vaa.clone(),
+                proof: merkle_acc
+                    .prove(m.as_ref())
+                    .ok_or(anyhow!("Failed to prove message"))?,
+            })
+        })
+        .collect::<Result<Vec<WormholeMerkleMessageProof>>>()
+}
+
+pub fn construct_update_data(mut message_states: Vec<MessageState>) -> Result<Vec<Vec<u8>>> {
+    message_states.sort_by_key(
+        |m| m.proof_set.wormhole_merkle_proof.vaa.clone(), // FIXME: This is not efficient
```

Comment:
there must be a way to do this without cloning

Reply:
yeah i will revise all the clones + other traits later as I mentioned in the PR description.